cpp/src/arrow/dataset/dataset_encryption_test.cc

```diff
@@ -0,0 +1,402 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <string_view>
+
+#include "gtest/gtest.h"
+
+#include "arrow/api.h"
+#include "arrow/array/builder_primitive.h"
+#include "arrow/builder.h"
+#include "arrow/dataset/api.h"
+#include "arrow/dataset/parquet_encryption_config.h"
+#include "arrow/dataset/partition.h"
+#include "arrow/filesystem/mockfs.h"
+#include "arrow/io/api.h"
+#include "arrow/status.h"
+#include "arrow/table.h"
+#include "arrow/testing/gtest_util.h"
+#include "parquet/arrow/reader.h"
+#include "parquet/encryption/test_in_memory_kms.h"
+
+constexpr std::string_view kFooterKeyMasterKey = "0123456789012345";
+constexpr std::string_view kFooterKeyMasterKeyId = "footer_key";
+constexpr std::string_view kColumnMasterKeys[] = {"1234567890123450"};
+constexpr std::string_view kColumnMasterKeysIds[] = {"col_key"};
+constexpr std::string_view kBaseDir = "";
+const int kNumColumns = 1;
+
+namespace arrow {
+namespace dataset {
+
+class DatasetEncryptionTest : public ::testing::Test {
+ protected:
+  // Create our parquetfileformat with encryption properties
+  std::pair<std::shared_ptr<DatasetEncryptionConfiguration>,
+            std::shared_ptr<DatasetDecryptionConfiguration>>
+  CreateDatasetEncryptionConfig(const std::string_view* column_ids,
+                                const std::string_view* column_keys, int num_columns,
+                                std::string_view footer_id, std::string_view footer_key,
+                                std::string_view footer_key_name = "footer_key",
+                                std::string_view column_key_mapping = "col_key: a") {
+    auto key_list =
+        BuildKeyMap(column_ids, column_keys, num_columns, footer_id, footer_key);
+
+    auto crypto_factory = SetupCryptoFactory(/*wrap_locally=*/true, key_list);
+
+    auto encryption_config =
+        std::make_shared<::parquet::encryption::EncryptionConfiguration>(
+            std::string(footer_key_name));
+    encryption_config->column_keys = column_key_mapping;
+    if (footer_key_name.size() > 0) {
+      encryption_config->footer_key = footer_key_name;
+    }
+
+    // DatasetEncryptionConfiguration
+    DatasetEncryptionConfiguration dataset_encryption_config;
```

Comment:
It would be good to call Setup() function here and below. And do not use default config implicitly.

Reply:
BTW, why not directly do this
```cpp
auto dataset_encryption_config = std::make_shared<DatasetEncryptionConfiguration>();
dataset_encryption_config->Setup(crypto_factory, connection_config, encryption_config);

auto dataset_decryption_config = std::make_shared<DatasetDecryptionConfiguration>();
dataset_decryption_config->Setup(crypto_factory, connection_config, decryption_config);

return std::make_pair(std::move(dataset_encryption_config), std::move(dataset_decryption_config));
```